manager/controllers/app/m4dapplication_controller_unit_test.go

```diff
@@ -177,3 +159,640 @@ func readApplication(f string) (*app.M4DApplication, error) {
 	}
 	return application, nil
 }
+
+func createM4DApplication(objectKey types.NamespacedName, n int) *app.M4DApplication {
+	labels := map[string]string{"app": "workload"}
+	return &app.M4DApplication{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      objectKey.Name,
+			Namespace: objectKey.Namespace,
+		},
+		Spec: app.M4DApplicationSpec{
+			Selector: app.Selector{ClusterName: "thegreendragon", WorkloadSelector: metav1.LabelSelector{MatchLabels: labels}},
+			AppInfo:  map[string]string{"intent": "Testing"},
+			Data:     make([]app.DataContext, n),
+		},
+	}
+}
+
+// create a storage account and a secret with credentials for this account
+func createStorageAccount(cl client.Client, region string) error {
+	// Create storage account
+	accountName := "account" + utils.Hash(region, 10)
+	secretName := "secret" + utils.Hash(region, 10)
+	account := &app.M4DStorageAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      accountName,
+			Namespace: utils.GetSystemNamespace(),
+		},
+		Spec: app.M4DStorageAccountSpec{
+			Endpoint:  "http://endpoint1",
+			SecretRef: secretName,
+			Regions:   []string{region},
+		},
+	}
+
+	dummySecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      secretName,
+			Namespace: utils.GetSystemNamespace(),
+		},
+		Data: map[string][]byte{"accessKeyID": []byte("value1"), "secretAccessKey": []byte("value2")},
+		Type: "Opaque",
+	}
+	if err := cl.Create(context.Background(), dummySecret); err != nil {
+		return err
+	}
+	return cl.Create(context.Background(), account)
+}
+
+func createReadModule(api *app.InterfaceDetails, source *app.InterfaceDetails) *app.M4DModule {
+	return &app.M4DModule{
```

Comment:
I think it would be better to have all the CRD definitions and K8s objects here as yamls in the testdata folder. That way one could manually play around with the files while developing and possibly deploy them as examples to kind or similar.

Reply:
Added yamls in unittests folder.